### PR TITLE
Add backoff to retry policy

### PR DESCRIPTION
**Pull request by zhexuany**
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tikv-client-lib-java/104)

### CODE REVIEW

**Comment by zhexuany** [comment]
@ilovesoup @birdstorm PTAL

**Review by birdstorm — APPROVED** [review]
LGTM

**Comment by birdstorm** [comment]
maybe we should check if fib overflows integer

File: src/main/java/com/pingcap/tikv/policy/RetryNTimes.java [hunk 1]
```diff
@@ -18,34 +18,44 @@
...
     super(handler);
     Preconditions.checkArgument(n >= 1, "Retry count cannot be less than 1.");
-    this.n = n;
+    this.fibonacci = generateFibonacci(n);
   }
 
   @Override
-  protected boolean shouldRetry(Exception e) {
-    return --n != 0;
+  public int[] getFibonacci() {
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/policy/RetryNTimes.java]: If backoff strategy is implemented in parent class, please put this in parent.
If you decide to make it in subclass, you might move callWithRetry here.
A single logic is better be contained with in a single class for clear design.

Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/policy/RetryNTimes.java]: I am rewriting it. Original design is not so good.

File: src/main/java/com/pingcap/tikv/policy/RetryPolicy.java [hunk 2]
```diff
@@ -36,47 +39,53 @@
...
+    doWait(millis);
   }
 
-  protected boolean checkNotRecoverableException(Status status) {
-    return unrecoverableStatus.contains(status.getCode());
+  private void doWait(long millis) {
+    try {
+      Thread.sleep( backOff.nextBackOffMillis() );
+    } catch (InterruptedException e) {
+      throw new RuntimeException( e );
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/policy/RetryPolicy.java]: Use a specific runtime error. All exceptions we defined are runtime exception error.

File: src/main/java/com/pingcap/tikv/policy/RetryNTimes.java [hunk 3]
```diff
@@ -17,35 +17,29 @@
...
+import com.pingcap.tikv.util.ExponentialBackOff;
 
-public class RetryNTimes extends RetryPolicy {
-  private int n;
-
-  private RetryNTimes(int n, ErrorHandler handler) {
+public class RetryNTimes<T> extends RetryPolicy<T> {
+  private RetryNTimes(int n, ErrorHandler<T> handler) {
     super(handler);
+    this.backOff = new ExponentialBackOff(3);
```
Comment (ilovesoup) [inline src/main/java/com/pingcap/tikv/policy/RetryNTimes.java]: If backoff is decoupled, this should be decoupled here as well. Lift this up to PolicyBuilder

**Review by ilovesoup — APPROVED** [review]
LGMT as long as comment addressed.

File: src/main/java/com/pingcap/tikv/policy/RetryPolicy.java (hunk 2 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/policy/RetryPolicy.java]: DONE

File: src/main/java/com/pingcap/tikv/policy/RetryNTimes.java (hunk 3 above)
Comment (zhexuany) [reply src/main/java/com/pingcap/tikv/policy/RetryNTimes.java]: Done.

**Pull request merged**